tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/MetricsHelper.java

```diff
@@ -0,0 +1,289 @@
+
+    Operand<TBool> validNonsclar =
+        canBroadcastNonscalarShapes(tf, weightsRank, weightsShape, valuesRank, valuesShape);
+
+    Operand<TBool> isValidShape = tf.select(isScalar, isScalar, validNonsclar);
+
+    return tf.withSubScope("broadcastWeights-dynamic").assertThat(isValidShape, data);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have the same rank and valid dimensions.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsRank the operand for the rank of the sample weights
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesRank the operand for the rank of the sample weights
+   * @param valuesShape the operand for the shape of the sample weights
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the Shape is scalar or not.
+   */
+  private static <T extends TNumber> Operand<TBool> canBroadcastNonscalarShapes(
+      Ops tf,
+      Operand<T> weightsRank,
+      Operand<T> weightsShape,
+      Operand<T> valuesRank,
+      Operand<T> valuesShape) {
+    tf = tf.withSubScope("hasValidNonscalarShape");
+    Operand<TBool> isSameRank = tf.math.equal(valuesRank, weightsRank);
+    return tf.select(isSameRank, canBroadcastDims(tf, weightsShape, valuesShape), isSameRank);
+  }
+
+  /**
+   * Gets an operand that tests if the shapes have valid dimensions or not.
+   *
+   * @param tf the TensorFlow Ops
+   * @param weightsShape the operand for the shape of the sample weights
+   * @param valuesShape the operand for the shape of the values
+   * @param <T> the data type for the operands
+   * @return a boolean operand to determine if the shapes have valid dimensions or not.
+   */
+  private static <T extends TNumber> Operand<TBool> canBroadcastDims(
+      Ops tf, Operand<T> weightsShape, Operand<T> valuesShape) {
+    tf = tf.withSubScope("hasValidDims");
+    Operand<T> valuesShape2d = tf.expandDims(valuesShape, tf.constant(-1));
+    Operand<T> validDims =
+        tf.concat(Arrays.asList(valuesShape2d, tf.onesLike(valuesShape2d)), tf.constant(1));
+    SetDiff1d<T, TInt32> invalidDimsDiff = tf.setDiff1d(weightsShape, tf.shape.flatten(validDims));
+    Operand<T> invalidDims = invalidDimsDiff.out();
+    Operand<TInt32> numInvalidDims = tf.size(invalidDims);
+    return tf.math.equal(tf.constant(0), numInvalidDims);
+  }
+
+  /**
+   * Broadcast `weights` to the same shape as `values`.
+   *
+   * @param tf the TensorFlow ops
+   * @param weights `Tensor` whose shape is broadcastable to `values`
+   * @param values Tensor` of any shape
+   * @param <T> the type of Operands
+   * @return <code>weights</code> broadcast to <code>values</code> shape
+   */
+  public static <T extends TNumber> Operand<T> broadcastWeights(
+      Ops tf, Operand<T> weights, Operand<T> values) {
+
+    Shape weightsShape = weights.shape();
+    Shape valuesShape = values.shape();
+
+    if (!weightsShape.hasUnknownDimension()
```

Comment:
FWIW, I'm pretty sure this logic is equivalent to 
~~~ Java
  if (!weightsShape.hasUnknownDimensions() && weightsShape.equals(valuesShape)) {
  . . .
  }
~~~

Reply:
The latest code has
```
if (!weightsShape.hasUnknownDimension()
        && !valuesShape.hasUnknownDimension()
        && weightsShape.isCompatibleWith(valuesShape)) {
      return weights;
    }
```
`weightsShape.isCompatibleWith(valuesShape)` is not the same as `weightsShape.equals(valuesShape)`

Reply:
👍 nothing to do